### ISSUE TITLE
Functional/orthology analysis

### ISSUE THREAD

**Issue by botellaflotante** [metagenome-atlas/atlas#330]
if I wanted to analyze KO / KEGG pathways or COGs between MAGs from different environments, what would you recommend to do with the eggNog annotations from ATLAS? as there are many KO numbers/COGs, which one should I use? do you reccomend any R package for this?

**Comment by SilasK** [comment]
I suggest you first to run this extension on atlas. 
https://github.com/metagenome-atlas/atlas_analyze

It produces:sparkles:[such cool report](http://htmlpreview.github.io/?https://github.com/metagenome-atlas/Tutorial/blob/master/Example/Results/Summary.html):sparkles: 

Which among others calculate the KO per sample based on the MAGs present.

There is python code to reproduce the report or if you prefer R. 
Here is the [R version](https://github.com/metagenome-atlas/Tutorial/blob/master/R/)
Setup can be done via the conda environment. 


Have also a look for the [interactive version](http://mybinder.org/v2/gh/metagenome-atlas/Tutorial/master?urlpath=rstudio) on a test dataset.  

If with your MAgs you have only a little coverage you might also want to quantify the [genes directly](https://github.com/metagenome-atlas/atlas/issues/276).

**Comment by botellaflotante** [comment]
Nice! Is it possible to run it if I have multiple samples but run separately in ATLAS?

**Comment by SilasK** [comment]
The short answer is no. 



I suggest, you to combine the different projects. This allows a consistent quantification of all samples on the same MAGs/genecatalog. Merging shouldn't be complicated: It's just coping the sample folders into a new folder with a samples.tsv describing all samples. Copy also the reports/stats/log folder in the new working directory. In this way, the assembly and binning for each sample is kept and only the steps of genecatalog and genome are rerun. 


If you don't want to combine the atlas projects: You can run atlas_analyze on all of them. You should get a relative abundance of genomes and functions for each of them. Then you can combine the tables (This you have to do yourselves)

**Comment by SilasK** [comment]
I haven't yet statistical test for comparing functions only genome abundances. But I would simply take a non-parametrical test e.g. mann-whitney and then do a correction for multiple testing. PCA of log2(psedocaount+relative abundance ) could also be an idea.  If you have other Ideas how to compare the functions between different samples I'm eager to hear.

**Comment by botellaflotante** [comment]
ok! to combine them, in the case of reports/stats/log folders, should I copy only one of them (any) in the new working dir?

**Comment by SilasK** [comment]
Yes, copy any of them.

**Comment by botellaflotante** [comment]
I got this error, has it happen to you before?
many thanks
output: Results/Summary.html
    shell:
        jupyter nbconvert --output Summary --TemplateExporter.exclude_input=True Results/Code.ipynb
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

**Comment by SilasK** [comment]
No, sorry this is not thorough fully tested. Are the no other stout?
Did you set up the atlas_analyze environment as described in the Readme of atlas_analyze?

What happens when you run 'jupyter nbconvert’ Maybe you need to install something more?

**Comment by SilasK** [comment]
I think this should solve your problem https://github.com/metagenome-atlas/atlas_analyze/issues/1

**Comment by botellaflotante** [comment]
Thanks! I already used the scripts without the nbconvert, very usefull!

> I suggest, you to combine the different projects. This allows a consistent quantification of all samples on the same MAGs/genecatalog. Merging shouldn't be complicated: It's just coping the sample folders into a new folder with a samples.tsv describing all samples. Copy also the reports/stats/log folder in the new working directory. In this way, the assembly and binning for each sample is kept and only the steps of genecatalog and genome are rerun.

When I copy the samples directories to a new one and add them the reports/stats/log folder together with config and samples files I get this error when trying to do the genomes step:

Building DAG of jobs...
CyclicGraphException in line 338 of /home/miniconda3/envs/atlasenv/lib/python3.6/site-packages/atlas/rules/genomes.smk:
Cyclic dependency on rule bam_2_sam_MAGs.

this is what I got in he dir
config.yaml  reports      SRR6797243  SRR6797246
logs         samples.tsv  SRR6797244  stats

I changed the samples.tsv to these 3 samples ...

best

**Comment by SilasK** [comment]
I know this an issue, I should correct. 

What happens if you set 'atlas run genecatalog' or 'atlas run genomes' ?

**Comment by botellaflotante** [comment]
Well this happens with any option (genomes, genecatalog, or all). I tried copying some bam files to genomes/alignment but did not work... I looked into the smk genome file but I don`t find where the problem would be (input asks for this bam file.. but..)
best

**Comment by botellaflotante** [comment]
I think this may be useful to add some more samples to a project, or also to compare different samples which were run independently before, without having to run everything from 0, right?

**Comment by botellaflotante** [comment]
> I haven't yet statistical test for comparing functions only genome abundances. But I would simply take a non-parametrical test e.g. mann-whitney and then do a correction for multiple testing. PCA of log2(psedocaount+relative abundance ) could also be an idea. If you have other Ideas how to compare the functions between different samples I'm eager to hear.

So, would it be possible to use ALDEx2 or something similar to do differential abundance of MAGs clr raw counts between samples? or would you use relative abundances for that?

**Comment by SilasK** [comment]
First of all sorry for the error about the cyclic dependency. 
> Cyclic dependency on rule bam_2_sam_MAGs.

Is this still a blocking issue, factor for you? My plan would be to drop the storage of bam, which removes the cyclic dependency. 

By any chance @jmtsuji did you encounter the cyclic dependency? 

For your question about statistics: HAve a look at the brand new Tutorial, where I show how to use Aldex for the genome abundance. https://github.com/metagenome-atlas/Tutorial

The abundance of function is based on the **relative abundance of genomes**, therefore it's better to use non-parametrical test instead of aldex. 

Kind regards
Silas

**Comment by botellaflotante** [comment]
> First of all sorry for the error about the cyclic dependency.
> 
> > Cyclic dependency on rule bam_2_sam_MAGs.
> 
> Is this still a blocking issue, factor for you? My plan would be to drop the storage of bam, which removes the cyclic dependency.
> 
> By any chance @jmtsuji did you encounter the cyclic dependency?
> 
> For your question about statistics: HAve a look at the brand new Tutorial, where I show how to use Aldex for the genome abundance. https://github.com/metagenome-atlas/Tutorial
> 
> The abundance of function is based on the **relative abundance of genomes**, therefore it's better to use non-parametrical test instead of aldex.
> 
> Kind regards
> Silas

Thanks Silas. Well, I am running again the samples from 0. But I think it could be good to have the chance to include new samples to a project that has already been run. If you lead me I can try do it...

I will check the tutorial, thanks. I guess that is what I was looking for, some way to assess if differences in MAGs abundance-compositions are significant or not... do you think this is aplicable to MAGs with a minimum completeness value, or for MAGs with 20% compl, would do it also?

Best

**Comment by jmtsuji** [comment]
@SilasK Interesting -- I've never run into that cyclic dependency error before. (Also, I regularly look through the BAM files output by ATLAS to link unassembled read analyses I do outside of ATLAS to genome bins produced by ATLAS, so I at least appreciate having access to them. :-)  )

**Comment by botellaflotante** [comment]
Do you know why is it better to use clr than count normalization from the DESeq package to analyze MAGs abundances across samples? this last should correct for seq depth and also differences in compositions from highly abundant MAGs changes, right?
Thanks

**Comment by SilasK** [comment]
I think I will point you to this article http://journal.frontiersin.org/article/10.3389/fmicb.2017.02224/full

At the end CLR and the DEseq normalisation are not completely different. It’s just that the CLR takes explicitly the compositional nature into account.

**Comment by botellaflotante** [comment]
ok thanks, and, would it be also possible to do a compositional approach for KO and CAZy analyses?

**Comment by SilasK** [comment]
I don’t know how to use a compositional approach for functions. If you have an idea I’m happy to discuss. The approach implemented in atlas, e.g. in the Tutorial. Sums the relative abundance of species that have a certain function. And the problem sums are no longer possible if you use log ratios.

An other approach would be to look for presence/absence but this is also not CoDA.

**Comment by botellaflotante** [comment]
yes, I don 't know if that's possible either. Do you know how to map these K numbers to level2/level3 functional categories in KEGG? is there any standard/easy way of doing it?
Thanks!

**Comment by SilasK** [comment]
I'm experimenting with DRAM #360
